abdiff/core/run_ab_transforms.py

```diff
@@ -1 +1,203 @@
+                    "AWS_SESSION_TOKEN": CONFIG.AWS_SESSION_TOKEN,
+                },
+                labels={
+                    "docker_image": docker_image,
+                    "source": source,
+                    "input_file": input_file,
+                    "run": run_id,
+                },
+                volumes=[f"{transformed_directory}:/tmp"],
+            )
+            logger.info(
+                f"Container '{container.id}' (Docker image: {docker_image}) "
+                f"RUNNING transform for '{source}' input_file: {input_file}."
+            )
+            container_ids.append(container.id)
+
+    # wait for all container runs to complete (status = "exited")
+    # get logs from all Docker containers
+    get_logs(run_id, container_ids, log_file=f"{run_directory}/transformed/logs.txt")
+
+    # clone run data and add list of input files
+    run_data = read_run_json(run_directory)
+    run_data.update({"input_files": input_files})
+    update_or_create_run_json(run_directory, run_data)
+
+    elapsed_time = perf_counter() - start_time
+    logger.info(
+        "Total time to complete process: %s", str(timedelta(seconds=elapsed_time))
+    )
+    return get_transformed_files(run_directory)
+
+
+def get_transformed_files(run_directory: str) -> tuple[list[str], ...]:
+    """Get list of filepaths to transformed JSON files.
+
+    Args:
+        run_directory (str): Run directory.
+
+    Returns:
+        tuple[list[str]]: Tuple containing lists of paths to transformed
+            JSON files for each image, relative to 'run_directory'.
+    """
+    filepaths = []
+    for root, _, files in os.walk(Path(run_directory) / "transformed"):
+        sub_filepaths = []
+        if root.endswith("transformed"):
+            # skip 'transformed' root directory
+            continue
+        for filename in files:
+            relative_dir = root.replace(run_directory, "")
+            sub_filepaths.append(os.path.join(relative_dir, filename))
+        filepaths.append(sub_filepaths)
+    return tuple(filepaths)
+
+
+def get_logs(run_id: str, container_ids: list[str], log_file: str) -> str:
+    """Monitor Docker container runs and write logs to file as containers exit."""
+    logged_container_ids: set[str] = set()
+    logs = ""
+    while logged_container_ids != set(container_ids):
+        if not DockerClient.containers.list(all=True, filters={"label": f"run={run_id}"}):
+            raise DockerContainersNotFoundError(run_id)
+
+        # get logs from completed container runs for a given "run"
+        exited_containers = [
+            container
+            for container in DockerClient.containers.list(
+                filters={"label": f"run={run_id}", "status": "exited"}
+            )
+            if container.id in container_ids and container.id not in logged_container_ids
+        ]
+        if any(exited_containers):
+            for container in exited_containers:
```

Comment:
If I'm understanding, this works nicely because we don't know how many containers finished since last we checked -- could be multiple -- but we sequentially write the logs out.  We avoid a pretty easy pitfall in parallel work where outputs come in from disparate tasks at the same time and jumble up logging or data collection.

Reply:
Indeed! When retrieving the "exited" containers, we are making sure to exclude exited containers that have already been _logged_ (i.e., `container.id not in logged_container_ids`) to avoid duplication and the structure of the for loop results in the logs writing to the file as they finish!